src/pool/srv_pool.c

```diff
@@ -1300,98 +1333,221 @@ ds_pool_enable_exclude(void)
+	ABT_mutex_lock(events->pse_mutex);
 
-	D_DEBUG(DB_MD, DF_UUID": queuing event: "DF_PS_EVENT"\n", DP_UUID(svc->ps_uuid),
-		DP_PS_EVENT(event));
+	if (events->pse_pending == NULL) {
+		rc = alloc_event_set(&events->pse_pending);
+		if (rc != 0)
+			goto out;
+		allocated = true;
+	}
+
+	rc = add_to_event_set(events->pse_pending, rank, incarnation, src, type);
+	if (rc != 0)
+		goto out;
+
+	events->pse_time = daos_gettime_coarse();
+
+	if (events->pse_paused) {
+		D_DEBUG(DB_MD, DF_UUID ": resuming event handling\n", DP_UUID(svc->ps_uuid));
+		events->pse_paused = false;
+	}
 
-	ABT_mutex_lock(events->pse_mutex);
-	d_list_add_tail(&event->psv_link, &events->pse_queue);
 	ABT_cond_broadcast(events->pse_cv);
+
+out:
+	if (rc != 0 && allocated)
+		free_event_set(&events->pse_pending);
 	ABT_mutex_unlock(events->pse_mutex);
-	return 0;
+	return rc;
 }
 
 static void
-discard_events(d_list_t *queue)
+resume_event_handling(struct pool_svc *svc)
 {
-	struct pool_svc_event  *event;
-	struct pool_svc_event  *tmp;
+	struct pool_svc_events *events = &svc->ps_events;
 
-	d_list_for_each_entry_safe(event, tmp, queue, psv_link) {
-		D_DEBUG(DB_MD, "discard event: "DF_PS_EVENT"\n", DP_PS_EVENT(event));
-		d_list_del_init(&event->psv_link);
-		D_FREE(event);
+	ABT_mutex_lock(events->pse_mutex);
+	if (events->pse_paused) {
+		D_DEBUG(DB_MD, DF_UUID ": resuming event handling\n", DP_UUID(svc->ps_uuid));
+		events->pse_paused = false;
+		ABT_cond_broadcast(events->pse_cv);
 	}
+	ABT_mutex_unlock(events->pse_mutex);
 }
 
-static int pool_svc_exclude_rank(struct pool_svc *svc, d_rank_t rank);
+static int pool_svc_exclude_ranks(struct pool_svc *svc, struct pool_svc_event_set *event_set);
 
-static void
-handle_event(struct pool_svc *svc, struct pool_svc_event *event)
+static int
+handle_event(struct pool_svc *svc, struct pool_svc_event_set *event_set)
 {
+	int i;
 	int rc;
 
-	if ((event->psv_src != CRT_EVS_GRPMOD && event->psv_src != CRT_EVS_SWIM) ||
-	    (event->psv_type == CRT_EVT_DEAD && pool_disable_exclude)) {
-		D_DEBUG(DB_MD, "ignore event: "DF_PS_EVENT" exclude=%d\n", DP_PS_EVENT(event),
-			pool_disable_exclude);
-		goto out;
+	D_INFO(DF_UUID ": handling event set: " DF_PS_EVENT_SET "\n", DP_UUID(svc->ps_uuid),
+	       DP_PS_EVENT_SET(event_set));
+
+	if (!pool_disable_exclude) {
+		rc = pool_svc_exclude_ranks(svc, event_set);
+		if (rc != 0) {
+			DL_ERROR(rc, DF_UUID ": failed to exclude ranks", DP_UUID(svc->ps_uuid));
+			return rc;
+		}
 	}
 
-	if (event->psv_rank == dss_self_rank() && event->psv_src == CRT_EVS_GRPMOD &&
-	    event->psv_type == CRT_EVT_DEAD) {
-		D_DEBUG(DB_MD, "ignore exclusion of self\n");
-		goto out;
+	/*
+	 * Check if the alive ranks are up in the pool map. If in the future we
+	 * add automatic reintegration below, for instance, we may need
+	 * to not only take svc->ps_lock, but also employ an RDB TX by
+	 * the book.
+	 */
+	ABT_rwlock_rdlock(svc->ps_pool->sp_lock);
+	for (i = 0; i < event_set->pss_len; i++) {
```

Comment:
no change requested, food for thought. If the struct pool_svc_event_set also kept track of how many of the events (out of pss_len total) are "alive" events, then maybe some optimizations are possible? For example skipping this lock/loop/unlock logic if number of alive events == 0?

Or in the opposite scenario (all events are "alive" events) then pool_svc_exclude_ranks wouldn't have to loop over pss_len items?

Reply:
Both are true, indeed. Because `event_set` is an array, not a linked list, I think (admittedly without benchmarking evidence) the looping over at most a couple of thousands of events is likely not a problem. Probably do the optimization if I turn out to be wrong, or when some new requirement also requires the tracking?

Reply:
Yes, sounds good to me.